### PR TITLE
Feature/add hashtable

### PR DESCRIPTION
**Pull request by louisashton**
Created a HashTable implementation

### CODE REVIEW

File: HashTable.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
+import java.util.*;
```
Comment (aidanns) [inline HashTable.java]: Use explicit imports rather than *.

**Comment by aidanns** [comment]
https://google.github.io/styleguide/javaguide.html is the style guide that I'll CR you against (roughly - some changes such as 120 character line length and using 4 spaces rather than 2)

File: HashTable.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+import java.lang.reflect.Array;
+
+class HashTable <K, V> {
```
Comment (aidanns) [inline HashTable.java]: Add a comment here in the following form:

```
/**
 * Simple sentence description of what this class does.
 * 
 * If needed, more paragraphs describing what the class does, why it exists and how it fits in.
 * 
 * @author Louis Ashton (louisashton@live.com)
 */
```

File: HashTable.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+    // The array cannot be smaller than MIN_CAPACITY at any time
```
Comment (aidanns) [inline HashTable.java]: Add a full stop to all comments to make them properly formed sentences.

File: HashTable.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    private static final int MIN_CAPACITY = 32;
```
Comment (aidanns) [inline HashTable.java]: Add a newline after the static variables.

File: HashTable.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+    private HashEntry[] table;
+
+    // The current number of elements.
+    private int size;
+
+    // The current capacity of the array.
+    private int capacity;
+
+    // initialLoad determines how full the array should be made when resizing
```
Comment (aidanns) [inline HashTable.java]: Add a full stop here.

File: HashTable.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+    private double initialLoad;
+    private double maxLoad;
+    private double minLoad;
+
+    public HashTable(){
```
Comment (aidanns) [inline HashTable.java]: Space between `()` and `{`

Comment (aidanns) [inline HashTable.java]: Add a comment for all constructors. Something simple like this when there are no args:

```
/**
 * Creates a new HashTable.
 */
```

File: HashTable.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        size = 0;
+        capacity = MIN_CAPACITY;
+        maxLoad = 0.75;
+        minLoad = 0.25;
+        this.initialLoad = 0.5;
+
+        //HashEntry[] table = new HashEntry[capacity];
```
Comment (aidanns) [inline HashTable.java]: Delete dead code.

File: HashTable.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        table = (HashEntry[]) Array.newInstance(HashEntry.class, capacity);
+    }
+
+    /**
+    * Returns the value associated with the specified key.
```
Comment (aidanns) [inline HashTable.java]: Make the `*`'s align with the first `*` on the first line.

File: HashTable.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    * @param key the key
```
Comment (aidanns) [inline HashTable.java]: Add a blank line with only the `*` between the first sentence and the `@param` entries.

File: HashTable.java [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    * @return the value associated with <tt>key</tt>;
```
Comment (aidanns) [inline HashTable.java]: Make this a proper sentence with a capitol letter and a full stop.

Comment (aidanns) [inline HashTable.java]: I wouldn't bother with the HTML tags here.

File: HashTable.java [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+    *         <tt>null</tt> if no such value
+    * @throws NullPointerException if <tt>key</tt> is <tt>null</tt>
+    */
+    public synchronized V get(K key){
+        if (key == null) throw new NullPointerException("first argument to get(K key) is null");
```
Comment (aidanns) [inline HashTable.java]: Always use braces for blocks:

```
if (key == null) {
   // Code goes here.
}
```

File: HashTable.java [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+        // Finds the key.
+        int index = hash(key) % capacity;
+        while(!key.equals(table[index].getKey())){
+            index = (index + 1) % capacity;
+        }
+        // Returns the value unless it is null.
+        return table[index]==null? null : table[index].getValue();
```
Comment (aidanns) [inline HashTable.java]: Add spaces around the operators. Structure it so the non-null case comes first to match your description.

```
return table[index] != null ? table[index].getValue() : null
```

File: HashTable.java [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+    */
+    public synchronized V get(K key){
+        if (key == null) throw new NullPointerException("first argument to get(K key) is null");
+
+        // Finds the key.
+        int index = hash(key) % capacity;
+        while(!key.equals(table[index].getKey())){
+            index = (index + 1) % capacity;
+        }
+        // Returns the value unless it is null.
```
Comment (aidanns) [inline HashTable.java]: This comment doesn't add much value, it's just saying what the line of code does - remove it.

File: HashTable.java [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+        return table[index]==null? null : table[index].getValue();
+    }
+
+    /**
+    * @param  key the key
```
Comment (aidanns) [inline HashTable.java]: Add a single sentence describing what the method does as the first line of the Javadoc comment.

Comment (aidanns) [inline HashTable.java]: Don't bother aligning the variable names in Javadoc.

Comment (aidanns) [inline HashTable.java]: Should be full sentences with capitol letters and full stops.

File: HashTable.java [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+    * @param  value the value
+    * @throws NullPointerException if <tt>key</tt> is <tt>null</tt>
+    */
+    public synchronized void put(K key, V value){
+        if (key == null) throw new NullPointerException("first argument to put(K key, ...) is null");
```
Comment (aidanns) [inline HashTable.java]: Curlies for blocks.

File: HashTable.java [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+        int index = hash(key) % capacity;
+        if (value == null) {
+	    throw new NullPointerException("second argument to put(..., V Value) is null");
+	}
```
Comment (aidanns) [inline HashTable.java]: Correct indentation.

File: HashTable.java [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+        while (table[index] != null && !key.equals(table[index].getKey())) {
+            index = (index + 1) % capacity;
+        }
+
+        //Increases the size if a new key is being put in.
```
Comment (aidanns) [inline HashTable.java]: Space after `//`

File: HashTable.java [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+        if (table[index] == null) 
+            size++;
+        table[index] = new HashEntry(key, value);
+
+        // Resizes if required.
+        resize();
```
Comment (aidanns) [inline HashTable.java]: Just change the method name to `resizeIfRequired()` and remove the comment.

File: HashTable.java [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+        //Increases the size if a new key is being put in.
+        if (table[index] == null) 
+            size++;
+        table[index] = new HashEntry(key, value);
+
+        // Resizes if required.
+        resize(); 
+    }
+
+    // Returns the absolute value of the hashed key
```
Comment (aidanns) [inline HashTable.java]: Javadoc comment style.

File: HashTable.java (hunk 2 above)
Comment (aidanns) [inline HashTable.java]: No space between the class name and the parametised types.

File: HashTable.java [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+    * @return the value associated with <tt>key</tt>;
+    *         <tt>null</tt> if no such value
+    * @throws NullPointerException if <tt>key</tt> is <tt>null</tt>
+    */
+    public synchronized V get(K key){
+        if (key == null) throw new NullPointerException("first argument to get(K key) is null");
+
+        // Finds the key.
+        int index = hash(key) % capacity;
+        while(!key.equals(table[index].getKey())){
```
Comment (aidanns) [inline HashTable.java]: Space between `)` and `{`

File: HashTable.java [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+        // Resizes if required.
+        resize(); 
+    }
+
+    // Returns the absolute value of the hashed key
+    private int hash(K key){
+        int hashValue = key.hashCode();
+        hashValue ^= (hashValue >>> 20) ^ (hashValue >>> 12);
+        hashValue = hashValue ^ (hashValue >>> 7) ^ (hashValue >>> 4);
+	return Math.abs(hashValue);
```
Comment (aidanns) [inline HashTable.java]: Consistent indentation.

File: HashTable.java [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+        table[index] = new HashEntry(key, value);
+
+        // Resizes if required.
+        resize(); 
+    }
+
+    // Returns the absolute value of the hashed key
+    private int hash(K key){
+        int hashValue = key.hashCode();
+        hashValue ^= (hashValue >>> 20) ^ (hashValue >>> 12);
```
Comment (aidanns) [inline HashTable.java]: Where did this algorithm come from out of interest?

File: HashTable.java [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+        hashValue = hashValue ^ (hashValue >>> 7) ^ (hashValue >>> 4);
+	return Math.abs(hashValue);
+    }
+
+    // Resizes the array when required.
```
Comment (aidanns) [inline HashTable.java]: Use correct Javadoc style comments.

File: HashTable.java [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+    private void resize(){
+        if(!((size < capacity*minLoad && capacity > MIN_CAPACITY) || size > capacity*maxLoad)){
```
Comment (aidanns) [inline HashTable.java]: Add a space between `)` and `{`.

File: HashTable.java [hunk 25]
```diff
@@ -0,0 +1,192 @@
...
+            return;
+        }
+        // The size of the new array
```
Comment (aidanns) [inline HashTable.java]: Remove this comment, doesn't add value.

File: HashTable.java (hunk 2 above)
Comment (aidanns) [inline HashTable.java]: `public class`

File: HashTable.java (hunk 3 above)
Comment (aidanns) [inline HashTable.java]: Don't think this comment adds value, remove it.

File: HashTable.java [hunk 26]
```diff
@@ -0,0 +1,192 @@
+import java.util.*;
+import java.lang.reflect.Array;
+
+class HashTable <K, V> {
+
+    // The array cannot be smaller than MIN_CAPACITY at any time
+    private static final int MIN_CAPACITY = 32;
+    private int approxSize = Integer.highestOneBit(MIN_CAPACITY) << 1;
```
Comment (aidanns) [inline HashTable.java]: I don't think this is used anywhere.

File: HashTable.java [hunk 27]
```diff
@@ -0,0 +1,192 @@
...
+    private int tableMask = approxSize - 1;
```
Comment (aidanns) [inline HashTable.java]: I don't think this is used anywhere.

File: HashTable.java [hunk 28]
```diff
@@ -0,0 +1,192 @@
...
+
+    private HashEntry[] table;
+
+    // The current number of elements.
+    private int size;
```
Comment (aidanns) [inline HashTable.java]: Change this variable name to `currentSize` and remove the comment.

File: HashTable.java [hunk 29]
```diff
@@ -0,0 +1,192 @@
...
+
+    // The current capacity of the array.
+    private int capacity;
```
Comment (aidanns) [inline HashTable.java]: Change this variable to `currentCapacity` and remove the comment.

File: HashTable.java [hunk 30]
```diff
@@ -0,0 +1,192 @@
...
+
+    // initialLoad determines how full the array should be made when resizing
+    private double initialLoad;
```
Comment (aidanns) [inline HashTable.java]: Change this variable to `desiredLoadAfterResize` and remove the comment.

File: HashTable.java (hunk 26 above)
Comment (aidanns) [inline HashTable.java]: Avoid setting the value for non static variables outside of the default constructor.

File: HashTable.java [hunk 31]
```diff
@@ -0,0 +1,192 @@
...
+    private double maxLoad;
+    private double minLoad;
+
+    public HashTable(){
+        size = 0;
+        capacity = MIN_CAPACITY;
+        maxLoad = 0.75;
+        minLoad = 0.25;
+        this.initialLoad = 0.5;
```
Comment (aidanns) [inline HashTable.java]: Don't use `this` unless it's needed.

File: HashTable.java (hunk 9 above)
Comment (aidanns) [inline HashTable.java]: Proper sentence here.

File: HashTable.java [hunk 32]
```diff
@@ -0,0 +1,192 @@
...
+        //HashEntry[] table = new HashEntry[capacity];
+        table = (HashEntry[]) Array.newInstance(HashEntry.class, capacity);
+    }
+
+    /**
+    * Returns the value associated with the specified key.
+    * @param key the key
+    * @return the value associated with <tt>key</tt>;
+    *         <tt>null</tt> if no such value
+    * @throws NullPointerException if <tt>key</tt> is <tt>null</tt>
```
Comment (aidanns) [inline HashTable.java]: Proper sentence here.

File: HashTable.java (hunk 15 above)
Comment (aidanns) [inline HashTable.java]: Use Google's Preconditions library for this kind of stuff: https://guava-libraries.googlecode.com/svn-history/r14/trunk/javadoc/com/google/common/base/Preconditions.html

File: HashTable.java [hunk 33]
```diff
@@ -0,0 +1,192 @@
...
+    private void resize(){
+        if(!((size < capacity*minLoad && capacity > MIN_CAPACITY) || size > capacity*maxLoad)){
+            return;
+        }
+        // The size of the new array
+        int newCapacity = (int) (size/initialLoad);
+
+        @SuppressWarnings("unchecked")
+        // Makes the new array
+        HashEntry[] newArray = (HashEntry[]) Array.newInstance(HashEntry.class, newCapacity);
```
Comment (aidanns) [inline HashTable.java]: Prefer

```
HashEntry[] newArray = new HashEntry[newCapacity];
```

File: HashTable.java [hunk 34]
```diff
@@ -0,0 +1,192 @@
...
+        //HashEntry[] newArray = new HashEntry[newCapacity];
+
+        for (int oldIndex = 0; oldIndex < capacity; oldIndex++){
```
Comment (aidanns) [inline HashTable.java]: Prefer `i` as the iterator location.

File: HashTable.java [hunk 35]
```diff
@@ -0,0 +1,192 @@
...
+            HashEntry oldEntry = table[oldIndex];
+            if (oldEntry == null)
```
Comment (aidanns) [inline HashTable.java]: Use curlies.

File: HashTable.java [hunk 36]
```diff
@@ -0,0 +1,192 @@
...
+                continue;
+
+            int index = hash(oldEntry.getKey()) % newCapacity;
+            while (newArray[index]!=null && !oldEntry.getKey().equals(newArray[index].getKey())) {
+                // get next index
+                index = (index + 1) % newCapacity;
+            }
+            newArray[index] = oldEntry;
+        }
+        this.table = newArray;
```
Comment (aidanns) [inline HashTable.java]: Don't use `this` when it's not needed.

File: HashTable.java [hunk 37]
```diff
@@ -0,0 +1,192 @@
...
+        this.capacity = newCapacity;
+    }
+
+    // Returns the number of elements in the hashtable.
```
Comment (aidanns) [inline HashTable.java]: Javadoc comment here.

File: HashTable.java [hunk 38]
```diff
@@ -0,0 +1,192 @@
...
+    public int size(){
```
Comment (aidanns) [inline HashTable.java]: Space between `)` and `{`.

File: HashTable.java [hunk 39]
```diff
@@ -0,0 +1,192 @@
...
+        return size;
+    }
+
+    // Returns all the keys in the hashtable.
```
Comment (aidanns) [inline HashTable.java]: Javadoc comment here.

File: HashTable.java [hunk 40]
```diff
@@ -0,0 +1,192 @@
...
+    public Set<K> getAll(){
```
Comment (aidanns) [inline HashTable.java]: Space between `)` and `{`.

File: HashTable.java [hunk 41]
```diff
@@ -0,0 +1,192 @@
...
+        Set<K> keys = new HashSet<K>(size);
+        for(HashEntry entry : table)
```
Comment (aidanns) [inline HashTable.java]: Use curlies.

File: HashTable.java [hunk 42]
```diff
@@ -0,0 +1,192 @@
...
+            if(entry != null)
+                keys.add(entry.getKey());
+        return keys;
+    }
+
+    // Removes the key from the hashtable.
```
Comment (aidanns) [inline HashTable.java]: Javadoc comment.

File: HashTable.java [hunk 43]
```diff
@@ -0,0 +1,192 @@
...
+    public Set<K> getAll(){
+        Set<K> keys = new HashSet<K>(size);
+        for(HashEntry entry : table)
+            if(entry != null)
+                keys.add(entry.getKey());
+        return keys;
+    }
+
+    // Removes the key from the hashtable.   
+    public void delete(K key){
```
Comment (aidanns) [inline HashTable.java]: Space between `)` and `{`.

File: HashTable.java [hunk 44]
```diff
@@ -0,0 +1,192 @@
...
+        if (key == null) throw new NullPointerException("first argument to put(K key) is null");
+        List<HashEntry> entries = new ArrayList<HashEntry>();
+
+        // Locates the key.
+        int index = hash(key) % capacity;
+        while (table[index]!=null && !key.equals(table[index].getKey())){
```
Comment (aidanns) [inline HashTable.java]: Spaces around operators.

File: HashTable.java [hunk 45]
```diff
@@ -0,0 +1,192 @@
...
+            index = (index+1) % capacity;
```
Comment (aidanns) [inline HashTable.java]: Spaces around operators.

File: HashTable.java [hunk 46]
```diff
@@ -0,0 +1,192 @@
...
+
+        // Extracts keys that collided with this key.
+        while(table[index] != null){
+            entries.add(table[index]);
+            table[index] = null;
+            size--;
+            index = (index+1) % capacity;
+        }
+
+        entries.remove(0); // Ignore the key to be deleted.
```
Comment (aidanns) [inline HashTable.java]: Comments always go on their own line, above the thing they are commenting.

File: HashTable.java [hunk 47]
```diff
@@ -0,0 +1,192 @@
...
+
+        for (HashEntry entry : entries)
```
Comment (aidanns) [inline HashTable.java]: Curlies.

File: HashTable.java [hunk 48]
```diff
@@ -0,0 +1,192 @@
...
+            this.put(entry.getKey(), entry.getValue()); // Puts the rest back in the hashtable, so that the linear probe is maintained
```
Comment (aidanns) [inline HashTable.java]: Comment on it's own line.

File: HashTable.java [hunk 49]
```diff
@@ -0,0 +1,192 @@
...
+            if (table[index] == null)
+                System.out.printf("Key %s already deleted %n", key.toString());
+        }
+
+        // Extracts keys that collided with this key.
+        while(table[index] != null){
+            entries.add(table[index]);
+            table[index] = null;
+            size--;
+            index = (index+1) % capacity;
```
Comment (aidanns) [inline HashTable.java]: Spaces around operators.

File: HashTable.java [hunk 50]
```diff
@@ -0,0 +1,192 @@
...
+        }
+
+        entries.remove(0); // Ignore the key to be deleted.
+
+        for (HashEntry entry : entries)
+            this.put(entry.getKey(), entry.getValue()); // Puts the rest back in the hashtable, so that the linear probe is maintained
+    }
+
+    public String toString(){
```
Comment (aidanns) [inline HashTable.java]:
```
@Override
```

File: HashTable.java [hunk 51]
```diff
@@ -0,0 +1,192 @@
...
+        return String.format("Hashtable(%f, %f, %f)", maxLoad, minLoad, initialLoad);
+    }
+
+    private class HashEntry {
+        private final K key;
```
Comment (aidanns) [inline HashTable.java]: newline.

File: HashTable.java [hunk 52]
```diff
@@ -0,0 +1,192 @@
...
+
+        for (HashEntry entry : entries)
+            this.put(entry.getKey(), entry.getValue()); // Puts the rest back in the hashtable, so that the linear probe is maintained
+    }
+
+    public String toString(){
+        return String.format("Hashtable(%f, %f, %f)", maxLoad, minLoad, initialLoad);
+    }
+
+    private class HashEntry {
```
Comment (aidanns) [inline HashTable.java]: Comment.

Comment (aidanns) [inline HashTable.java]: Add the inner class at the top of the file.

File: HashTable.java [hunk 53]
```diff
@@ -0,0 +1,192 @@
...
+        private final K key;
+        private final V value;
+
+        HashEntry(K key, V value) {
+            this.key = key;
+            this.value = value;
+        }     
+
+        public K getKey() {
```
Comment (aidanns) [inline HashTable.java]: Comment.

File: HashTable.java [hunk 54]
```diff
@@ -0,0 +1,192 @@
...
+
+    public String toString(){
+        return String.format("Hashtable(%f, %f, %f)", maxLoad, minLoad, initialLoad);
+    }
+
+    private class HashEntry {
+        private final K key;
+        private final V value;
+
+        HashEntry(K key, V value) {
```
Comment (aidanns) [inline HashTable.java]: Comment.

File: HashTable.java [hunk 55]
```diff
@@ -0,0 +1,192 @@
...
+            this.key = key;
+            this.value = value;
+        }     
+
+        public K getKey() {
+            return key;
+        }
+
+        public V getValue() {
```
Comment (aidanns) [inline HashTable.java]: Comment.

File: HashTable.java [hunk 56]
```diff
@@ -0,0 +1,192 @@
...
+            return value;
+        }
+    }
+
+    public static void main(String arg[]) {
```
Comment (aidanns) [inline HashTable.java]: Move this in to a separate file.

Comment (aidanns) [inline HashTable.java]: Add some JUnit tests for this class.

**Comment by aidanns** [comment]
Let's change this project to use Gradle as a build system: https://gradle.org/getting-started-gradle-java/

**Comment by aidanns** [comment]
Let me know when you've had a chance to work through this and I'll take another look.

File: HashTable.java (hunk 22 above)
Comment (louisashton) [inline HashTable.java]: [http://stackoverflow.com/questions/2538092/why-does-a-hashmap-rehash-the-hashcode-supplied-by-the-key-object](url)

Comment (louisashton) [inline HashTable.java]: [From the java HashMap](http://stackoverflow.com/questions/2538092/why-does-a-hashmap-rehash-the-hashcode-supplied-by-the-key-object)

**Pull request closed without merging**